test/functional/api/v2/contents_controller_test.rb

```diff
@@ -0,0 +1,227 @@
+
+      assert_recognizes expected, "/api/v2/rubygems/foo/versions/1.0.0/contents.json"
+    end
+
+    should "route to index with .yaml" do
+      expected = { controller: "api/v2/contents", action: "index", rubygem_name: "foo", version_number: "1.0.0", format: "yaml" }
+
+      assert_recognizes expected, "/api/v2/rubygems/foo/versions/1.0.0/contents.yaml"
+    end
+
+    should "route to index with .sha256" do
+      expected = { controller: "api/v2/contents", action: "index", rubygem_name: "foo", version_number: "1.0.0", format: "sha256" }
+
+      assert_recognizes expected, "/api/v2/rubygems/foo/versions/1.0.0/contents.sha256"
+    end
+
+    should "not be confused by prerelease versions" do
+      expected = { controller: "api/v2/contents", action: "index", rubygem_name: "foo", version_number: "1.0.0-a.pre" }
+
+      assert_recognizes expected, "/api/v2/rubygems/foo/versions/1.0.0-a.pre/contents"
+    end
+
+    should "not route when disallowed characters are used" do
+      assert_raises(ActionController::UrlGenerationError) do
+        get :index, params: { rumgem_name: "foo", version_number: "bad%20version", format: "json" }
+      end
+    end
+  end
+
+  context "on GET to index" do
+    setup do
+      @rubygem = create(:rubygem)
+      @jruby_version = create(:version, rubygem: @rubygem, number: "2.0.0", platform: "jruby")
+      @version = create(:version, rubygem: @rubygem, number: "2.0.0")
+      create(:version, rubygem: @rubygem, number: "1.0.0.pre", prerelease: true)
+      create(:version, rubygem: @rubygem, number: "3.0.0", indexed: false)
+
+      @rubygem2 = create(:rubygem)
+      create(:version, rubygem: @rubygem2, number: "3.0.0")
+      create(:version, rubygem: @rubygem2, number: "2.0.0")
+      create(:version, rubygem: @rubygem2, number: "1.0.0")
+
+      @checksums = { "file.rb" => "file.rb-hash", "file2.rb" => "file2.rb-hash" }
+      @version.manifest.store_checksums(@checksums)
+
+      @jruby_checksums = { "file.rb" => "file.rb-jruby", "file2.rb" => "file2.rb-jruby" }
+      @jruby_version.manifest.store_checksums(@jruby_checksums)
+    end
+
+    should_respond_to(:json) do |body|
+      JSON.parse(body)
+    end
+
+    should_respond_to(:yaml) do |body|
+      YAML.safe_load(body)
+    end
+
+    context "with .sha256 format" do
+      should "return not found when the hashed files are not available" do
+        get_index(@rubygem, "1.0.0.pre")
+
+        assert_response :not_found
+        assert_equal "Content is unavailable for this version.", @response.body
+      end
+
+      should "return not found when the gem is not indexed" do
+        get_index(@rubygem, "3.0.0")
+
+        assert_response :not_found
+        assert_equal "This version could not be found.", @response.body
+      end
+
+      should "return a list of contents for the first gem" do
+        get_index(@rubygem, "2.0.0")
+
+        assert_response :success
+        assert_equal ShasumFormat.generate(@checksums), @response.body
```

Comment:
:thinking: It would be easier to follow to actually assert to fixed value (string) to see what response is expected.

Reply:
I like that idea. will do.